query.go

```diff
@@ -0,0 +1,431 @@
+				urlQuery[param.ParameterName],
+				param.StructFieldName,
+				err.Error(),
+			))
+			continue
+		}
+
+		field.Set(reflect.ValueOf(decodedParam))
+	}
+
+	if len(errs.Errors()) == 0 {
+		return nil
+	}
+	return errs
+}
+
+// This ignores the case of parameter name in favor of the canonical format of
+// http.Header
+func (qm QueryMap) EncodeHeader(src interface{}, headers http.Header) error {
+	srcVal := reflect.ValueOf(src)
+
+	for _, p := range qm.Parameters {
+		fieldVal := srcVal.FieldByName(p.StructFieldName)
+
+		if fieldVal.IsZero() && p.OmitEmpty {
+			continue
+		}
+
+		sliVal, err := p.Mapper.Encode(fieldVal)
+		if err != nil {
+			return errors.New("error in encoding struct: " + err.Error())
+		}
+
+		// Not using .Set() because it only allows strings and not slices
+		headers[http.CanonicalHeaderKey(p.ParameterName)] = sliVal
+	}
+
+	return nil
+}
+
+func (qm QueryMap) DecodeHeader(headers http.Header, dst interface{}) error {
+	if reflect.ValueOf(dst).Elem().Type() != reflect.TypeOf(qm.UnderlyingType) {
+		return errors.New("attempting to decode into the wrong struct")
+	}
+
+	// First sanity check to ensure that the struct passed in matches
+	// the struct the QueryMap was designed to handle
+	if reflect.ValueOf(dst).Elem().Type() != reflect.TypeOf(qm.UnderlyingType) {
+		return fmt.Errorf("attempting to decode into mismatched struct: expected %s but got %s",
+			reflect.TypeOf(qm.UnderlyingType),
+			reflect.ValueOf(dst).Elem().Type(),
+		)
+	}
+
+	errs := &MultiValidationError{}
+	dstVal := reflect.ValueOf(dst).Elem()
+	for _, param := range qm.Parameters {
+		headerVal := headers[http.CanonicalHeaderKey(param.ParameterName)]
+		if len(headerVal) == 0 {
+			continue
+		}
+
+		field := dstVal.FieldByName(param.StructFieldName)
+		decodedHeader, err := param.Mapper.Decode(headerVal)
+		if err != nil {
+			errs.AddError(NewValidationError("error ocurred while reading value (%s) into param %s: %s",
+				headerVal,
+				param.StructFieldName,
+				err.Error(),
+			))
+			continue
+		}
+
+		field.Set(reflect.ValueOf(decodedHeader))
+	}
+
+	if len(errs.Errors()) == 0 {
+		return nil
+	}
+	return errs
+}
+
+// MappedParameter corresponds to each field in a specific struct,
+// it requires struct's name and the corresponding key value in the URL query
+type MappedParameter struct {
+	StructFieldName string
+	ParameterName   string
+	Mapper          QueryParameterMapper
+	OmitEmpty       bool
+}
+
+// QueryParameterMapper defines how url.Values value ([]string) and struct are to be
+// transformed into each other. It is from a slice of strings, reflecting the structure
+// of url.Values. These can be specified by their type (whichever struct the Parameter
+// mapper will be, and the restrictions defined on the type, defined by Validators slice
+// below)
+type QueryParameterMapper interface {
+	Encode(reflect.Value) ([]string, error)
+	Decode([]string) (interface{}, error)
```

Comment:
nit: you could probably use a variadic function for `Decode` and you wouldn't have to wrap each string in an array for `StrSliceQueryParameterMapper`

Reply:
omg thanks--that was a terrible line of code